libraries/src/Installer/InstallerScriptTrait.php

```diff
@@ -0,0 +1,362 @@
+     *
+     * @return  boolean  True on success
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function postflight(string $type, InstallerAdapter $adapter): bool
+    {
+        $this->removeFiles();
+
+        return $this->customPostflight($type, $adapter);
+    }
+
+    /**
+     * Custom postflight method to be overridden by the extension developer
+     *
+     * @param   string            $type     The type of change (install or discover_install, update, uninstall)
+     * @param   InstallerAdapter  $adapter  The adapter calling this method
+     *
+     * @return  boolean  True on success
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected function customPostflight(string $type, InstallerAdapter $adapter): bool
+    {
+        return true;
+    }
+
+    /**
+     * Check if the extension passes the minimum requirements to be installed
+     *
+     * @return  boolean  True on success
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected function checkCompatibility(string $type, InstallerAdapter $adapter): bool
+    {
+        // Check for the minimum PHP version before continuing
+        if (!empty($this->minimumPhp) && version_compare(PHP_VERSION, $this->minimumPhp, '<')) {
+            Log::add(Text::sprintf('JLIB_INSTALLER_MINIMUM_PHP', $this->minimumPhp), Log::WARNING, 'jerror');
+
+            return false;
+        }
+
+        // Check for the minimum Joomla version before continuing
+        if (!empty($this->minimumJoomla) && version_compare(JVERSION, $this->minimumJoomla, '<')) {
+            Log::add(Text::sprintf('JLIB_INSTALLER_MINIMUM_JOOMLA', $this->minimumJoomla), Log::WARNING, 'jerror');
+
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Check if the extension is allowed to be downgraded
+     *
+     * @return  boolean  False when downgrade not allowed and new version is lower than old version otherwise true
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected function checkDowngrade(string $type, InstallerAdapter $adapter): bool
+    {
+        if ($type !== 'update' || $this->allowDowngrades) {
+            return true;
+        }
+
+        $oldManifest = $this->getOldManifest($adapter);
+
+        if ($oldManifest !== null && $oldManifest->version && version_compare($oldManifest->version, $adapter->getManifest()->version, '>')) {
+            Log::add(Text::_('JLIB_INSTALLER_ERROR_DOWNGRADE'), Log::WARNING, 'jerror');
+
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns the manifest file if it exists or null
+     *
+     * @param InstallerAdapter $adapter
+     *
+     * @return SimpleXMLElement|null
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected function getOldManifest(InstallerAdapter $adapter): ?\SimpleXMLElement
+    {
+        $client = ApplicationHelper::getClientInfo(1);
```

Comment:
```suggestion
        $client = ApplicationHelper::getClientInfo('administrator', true);
```

no magic numbers please, also I think we should introduce a `ApplcationEnum` (out of scope for this PR)